app/src/main/java/org/simple/clinic/scheduleappointment/ScheduleAppointmentSheetController.kt

```diff
@@ -27,145 +34,219 @@ typealias UiChange = (Ui) -> Unit
-    val patientUuidStream = events
-        .ofType<ScheduleAppointmentSheetCreated>()
-        .map { it.patientUuid }
+  private fun previousConfiguredAppointmentDate(
+      latestAppointmentScheduledDate: PotentialAppointmentDate,
+      configuredAppointmentScheduledDates: List<PotentialAppointmentDate>
+  ): PotentialAppointmentDate {
+    return configuredAppointmentScheduledDates
+        .findLast { it < latestAppointmentScheduledDate }
+        ?: throw RuntimeException("Cannot find configured appointment date before ${latestAppointmentScheduledDate.scheduledFor}")
+  }
 
+  private fun scheduleAutomaticAppointmentForDefaulters(events: Observable<UiEvent>): Observable<UiChange> {
     val combinedStreams = Observables.combineLatest(
         events.ofType<SchedulingSkipped>(),
-        patientUuidStream,
-        configProvider.toObservable()
+        patientUuid(events),
+        configProvider
     )
     val isPatientDefaulterStream = combinedStreams
         .switchMap { (_, patientUuid) -> patientRepository.isPatientDefaulter(patientUuid) }
         .replay()
         .refCount()
 
-    val currentFacilityStream = userSession
-        .requireLoggedInUser()
-        .switchMap { user -> facilityRepository.currentFacility(user) }
-
     val saveAppointmentAndCloseSheet = isPatientDefaulterStream
         .filter { isPatientDefaulter -> isPatientDefaulter }
-        .withLatestFrom(patientUuidStream, configProvider.toObservable(), currentFacilityStream) { _, patientUuid, config, currentFacilty ->
+        .withLatestFrom(
+            patientUuid(events),
+            configProvider,
+            currentFacilityStream()
+        ) { _, patientUuid, config, currentFacilty ->
           Triple(patientUuid, config, currentFacilty)
         }
         .flatMapSingle { (patientUuid, config, currentFacility) ->
-          appointmentRepository
-              .schedule(
-                  patientUuid = patientUuid,
-                  appointmentDate = LocalDate.now(utcClock).plus(config.appointmentDuePeriodForDefaulters),
-                  appointmentType = Automatic,
-                  currentFacility = currentFacility)
-              .map { { ui: Ui -> ui.closeSheet() } }
+          scheduleAppointmentForPatient(
+              uuid = patientUuid,
+              date = LocalDate.now(clock).plus(config.appointmentDuePeriodForDefaulters),
+              currentFacility = currentFacility,
+              appointmentType = Appointment.AppointmentType.Automatic
+          )
         }
+        .map { Ui::closeSheet }
 
     val closeSheetWithoutSavingAppointment = isPatientDefaulterStream
         .filter { isPatientDefaulter -> isPatientDefaulter.not() }
-        .map { { ui: Ui -> ui.closeSheet() } }
+        .map { Ui::closeSheet }
 
     return Observable.merge(saveAppointmentAndCloseSheet, closeSheetWithoutSavingAppointment)
   }
 
   private fun scheduleCreates(events: Observable<UiEvent>): Observable<UiChange> {
```

Comment:
This function is scheduling the appointment and no more handling creation event. Let's call it `scheduleAppointmentDone`?

Reply:
Since this chain is still doing the job of actually creating the appointment, I have renamed it to `scheduleAppointment()`.